src/lib/tecs/TECS.cpp

```diff
@@ -41,241 +46,403 @@ using math::constrain;
 using math::max;
 using math::min;
 
-static constexpr float DT_MIN = 0.001f;	///< minimum allowed value of _dt (sec)
-static constexpr float DT_MAX = 1.0f;	///< max value of _dt allowed before a filter state reset is performed (sec)
+// TODO there seems to be an inconsistent definition of IAS/CAS/EAS/TAS
+// TODO Recheck the timing.
+void TECSAirspeedFilter::initialize(const float equivalent_airspeed)
+{
 
-/**
- * @file TECS.cpp
- *
- * @author Paul Riseborough
- */
 
-/*
- * This function implements a complementary filter to estimate the climb rate when
- * inertial nav data is not available. It also calculates a true airspeed derivative
- * which is used by the airspeed complimentary filter.
- */
-void TECS::update_vehicle_state_estimates(float equivalent_airspeed, const float speed_deriv_forward,
-		bool altitude_lock, float altitude, float vz)
+	_airspeed_state.speed = equivalent_airspeed;
+	_airspeed_state.speed_rate = 0.0f;
+	_airspeed_rate_filter.reset(0.0f);
+}
+
+void TECSAirspeedFilter::update(const float dt, const Input &input, const Param &param,
+				const bool airspeed_sensor_available)
 {
-	// calculate the time lapsed since the last update
-	uint64_t now = hrt_absolute_time();
-	float dt = fmaxf((now - _state_update_timestamp) * 1e-6f, DT_MIN);
+	// Input checking
+	if (!(PX4_ISFINITE(dt) && dt > FLT_EPSILON)) {
+		// Do not update the states.
+		PX4_WARN("Time intervall is not valid.");
+		return;
+	}
 
-	bool reset_altitude = false;
+	float airspeed;
 
-	if (_state_update_timestamp == 0 || dt > DT_MAX) {
-		dt = DT_DEFAULT;
-		reset_altitude = true;
-	}
+	if (PX4_ISFINITE(input.equivalent_airspeed) && airspeed_sensor_available) {
+		airspeed = input.equivalent_airspeed;
 
-	if (!altitude_lock) {
-		reset_altitude = true;
+	} else {
+		airspeed = param.equivalent_airspeed_trim;
 	}
 
-	if (reset_altitude) {
-		_states_initialized = false;
-	}
+	float airspeed_derivative;
 
-	_state_update_timestamp = now;
-	_EAS = equivalent_airspeed;
+	if (PX4_ISFINITE(input.equivalent_airspeed_rate) && airspeed_sensor_available) {
```

Comment:
@KonradRudin Actually this raises an interesting point. We could actually use this filter with synthetic airspeed. You'd get the benefit of a more dynamic airspeed (due to prediction using acceleration).

Reply:
I guess it depends on how good it is already and how much filtering needs to be done. But it would make sense to use the acceleration as well. I would leave the logic the same as before.